content/blog/2021/06/writing-your-first-policy-with-kubewarden.md

```diff
@@ -0,0 +1,553 @@
+			"resource": "namespaces"
+		},
+		"name": "test-policy-ns",
+		"namespace": "test-policy-ns",
+		"operation": "CREATE",
+		"userInfo": {
+			"username": "kubernetes-admin",
+			"groups": ["system:masters", "system:authenticated"]
+		},
+		"object": {
+			"kind": "Namespace",
+			"apiVersion": "v1",
+			"metadata": {
+				"name": "test-policy-ns",
+				"uid": "2cc6b1fc-7896-4091-b9c3-e034c89c0535",
+				"creationTimestamp": "2021-05-27T15:55:39Z",
+				"managedFields": [{
+					"manager": "kubectl",
+					"operation": "Update",
+					"apiVersion": "v1",
+					"time": "2021-05-27T15:55:39Z",
+					"fieldsType": "FieldsV1",
+					"fieldsV1": {
+						"f:status": {
+							"f:phase": {}
+						}
+					}
+				}]
+			},
+			"spec": {
+				"finalizers": ["kubernetes"]
+			},
+			"status": {
+				"phase": "Active"
+			}
+		},
+		"oldObject": null,
+		"dryRun": false,
+		"options": {
+			"kind": "CreateOptions",
+			"apiVersion": "meta.k8s.io/v1"
+		}
+	}
+}
+```
+
+I save this request as `request.json`. Now, we have all the
+ingredients to dry run the policy, so let's do it:
+
+```console
+$ kwctl run policy.wasm --request-path request.json 2> /dev/null | jq
+{
+  "uid": "150d1761-43fb-4f36-bb73-7a3888a0bca2",
+  "allowed": true
+}
+```
+
+As expected, the response of the policy is that the request is
+accepted. This is a really good start! Now, let's go and write the
+real policy logic.
+
+## Policy implementation
+
+We have to implement three main checks based on the restrictions that
+have been described by our compliance team:
+
+1. The namespace name has to meet a specific prefix format
+   (`t-\w+-\w+` -- e.g. `t-ei-billing`, `t-sre-security`...)
+1. Two annotations have to exist:
+    1. `compliance.my-company.com/team-contact-email`: has to be an
+    email address with the host part pointing to `@my-company.com`; we
+    won't allow other domain names as contact email.
+    1. `compliance.my-company.com/team-region`: has to be one (and
+only one) of `AMER`, `APAC` or `EMEA`.
+
+Let's implement the first check:
+
+```go
+package main
+
+import (
+	"fmt"
+	"github.com/kubewarden/gjson"
+	kubewarden "github.com/kubewarden/policy-sdk-go"
+	"regexp"
+	wapc "github.com/wapc/wapc-guest-tinygo"
+)
+
+func main() {
+	wapc.RegisterFunctions(wapc.Functions{
+		"validate": func(payload []byte) ([]byte, error) {
+			namespaceName := gjson.Get(string(payload), "request.object.metadata.name")
+			namespaceNameMatches, err := regexp.MatchString(`^t-\w+-\w+`, namespaceName.String())
+			if err != nil {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("internal error while processing namespace: %v", err)), kubewarden.NoCode)
+			}
+			if !namespaceNameMatches {
+				return kubewarden.RejectRequest(kubewarden.Message(fmt.Sprintf("namespace %s does not comply with the company namespace naming convention: check https://internal.my-company.com/k8s/naming-conventions.md for more information", namespaceName)), kubewarden.NoCode)
```

Comment:
nitpicking: I would surround the name of the invalid namespace with `'`

Reply:
Interestingly, using `"` or `'` broke something on our end so I went back to unquoted names so I could debug what's wrong later.

Reply:
Quoting with `'` works. Quoting with `"` wreaks havoc. We must be having an issue with json quoting.

```
ereslibre@desktop ~/sample (main)> kwctl run policy.wasm --request-path request.json
May 28 16:44:12.702 ERROR validate{self=PolicyEvaluator { settings: None } request="{\"uid\":\"150d1761-43fb-4f36-bb73-7a3888a0bca2\",\"kind\":{\"group\":\"\",\"version\":\"v1\",\"kind\":\"Namespace\"},\"resource\":{\"group\":\"\",\"version\":\"v1\",\"resource\":\"namespaces\"},\"requestKind\":{\"group\":\"\",\"version\":\"v1\",\"kind\":\"Namespace\"},\"requestResource\":{\"group\":\"\",\"version\":\"v1\",\"resource\":\"namespaces\"},\"name\":\"test-policy-ns\",\"namespace\":\"test-policy-ns\",\"operation\":\"CREATE\",\"userInfo\":{\"username\":\"kubernetes-admin\",\"groups\":[\"system:masters\",\"system:authenticated\"]},\"object\":{\"kind\":\"Namespace\",\"apiVersion\":\"v1\",\"metadata\":{\"name\":\"test-policy-ns\",\"uid\":\"2cc6b1fc-7896-4091-b9c3-e034c89c0535\",\"creationTimestamp\":\"2021-05-27T15:55:39Z\",\"managedFields\":[{\"manager\":\"kubectl\",\"operation\":\"Update\",\"apiVersion\":\"v1\",\"time\":\"2021-05-27T15:55:39Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:status\":{\"f:phase\":{}}}}]},\"spec\":{\"finalizers\":[\"kubernetes\"]},\"status\":{\"phase\":\"Active\"}},\"oldObject\":null,\"dryRun\":false,\"options\":{\"kind\":\"CreateOptions\",\"apiVersion\":\"meta.k8s.io/v1\"}}"}: policy_evaluator::policy_evaluator: cannot build validation response from policy result error="cannot deserialize policy validation response: Error(\"expected `,` or `}`\", line: 1, column: 41)"
```

I will have a look at this. JSON escaping can get noisy really fast... let's see and hope base64 is not the best way we have to share this information between host and guest.

Reply:
A clear winner: https://github.com/kubewarden/policy-sdk-go/blob/e3cd7155fc02f723ee101f0e280ce2cdf313d789/mod.go#L48

I'm single quoting and I will fix this on the SDK.